aten/src/ATen/native/CPUBlas.cpp

```diff
@@ -351,7 +352,15 @@ void gemm(
    at::Half *c, int64_t ldc) {
    internal::normalize_last_dims(transa, transb, m, n, k, &lda, &ldb, &ldc);
 #if AT_MKLDNN_ENABLED()
-   if (mkldnn_fp16_gemm(transa, transb, m, n, k, alpha, a, lda, b, ldb, beta, c, ldc)) {
+   // The documented MKL behavior when FP16 support is not available
+   // is to upconvert to fp32 and call sgemm. We can do better by
+   // fusing the conversion.
+   const bool fp16_gemv_trans_fast_path_would_be_beneficial =
+     cpuinfo_initialize() && cpuinfo_has_x86_f16c() && !cpuinfo_has_x86_avx512fp16();
```

Comment:
I guess checking `cpuinfo_has_x86_avx512fp16` is not necessary since onednn (mkldnn) won't use avx512fp16 to compute gemms by default because the avx512fp16 fma would incur accuracy loss.

Reply:
@jgong5 https://community.intel.com/t5/Intel-oneAPI-Math-Kernel-Library/FP16-GEMM-using-AVX512-on-Sapphire-Rapids/m-p/1570739 doesn't seem to agree. I'm also confused -- I thought FMA was supposed to improve accuracy because it has high internal precision, so the result of the multiply doesn't have to be rounded to FP16 before the addition.

Reply:
@swolchok The link you referred to is about MKL not oneDNN (or mkldnn). MKL has dedicated API (hgemm) that uses AVX512_FP16 instruction but users should be aware of the accuracy loss due to FP16 accumulators. It is not about the accumulator for a single FMA (which has high internal precision has you mentioned) but about accumulation along the K-dim across multiple FMAs. On the other hand, oneDNN uses FP32 accumulators to keep high accuracy.

Reply:
oh, we use FP32 accumulation as well.

Reply:
`mkldnn_fp16_gemm` despite the name is also available on ARM, it looks like your change will skip MKLDNN unless it is on x86 platform, wouldn't it? What is the motivation for it?

Reply:
> mkldnn_fp16_gemm despite the name is also available on ARM,

news to me!

> it looks like your change will skip MKLDNN unless it is on x86 platform, wouldn't it?

fortunately no, because ARM machines won't pass `cpuinfo_has_x86_f16c()`.

Reply:
> also available on ARM

I don't think so? https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/mkldnn/Utils.h#L120

something to look at for BF16 though.

Reply:
> oh, we use FP32 accumulation as well.

Will you remove this `cpuinfo_has_x86_avx512fp16()` check then? I don't think it is relevant.

Reply:
>  remove this cpuinfo_has_x86_avx512fp16() check then? I don't think it is relevant.

I am surprised, but you're definitely the authority on this and I don't have a Sapphire Rapids machine to test on. I'll leave a note for posterity though.